rpcbind_test.py fails for standard gitlab-ci configuration

From #3598


>It fails with 
>
>```
>AssertionError: not({('7f000001', 19180), ('00000000000000000000000000000001', 19180)} == {('00000000000000000000000000000000', 19180)})
>```
>here 
>https://github.com/dashpay/dash/blob/82472d1a706ddaed3ab7360ef1201fb2ff467c55/test/functional/rpcbind_test.py#L37
>
>due to this test 
>
>https://github.com/dashpay/dash/blob/82472d1a706ddaed3ab7360ef1201fb2ff467c55/test/functional/rpcbind_test.py#L81-L82
>
>It looks like 
>
>https://github.com/dashpay/dash/blob/82472d1a706ddaed3ab7360ef1201fb2ff467c55/test/functional/test_framework/netutil.py#L76-L85
>
>is just not working properly for some setups. Which causes this wrong "host,port" tuple -> `('7f000001', 19180)`
>
>and on top it looks like this 
>
>https://github.com/dashpay/dash/blob/82472d1a706ddaed3ab7360ef1201fb2ff467c55/test/functional/rpcbind_test.py#L81-L82
>
>should rather be 
>
>```
>self.run_bind_test(['127.0.0.1'], '127.0.0.1', [], 
>     [('::1', defaultport)]) 
>```
>
>Im not sure at all so this would require some investigation from my side.. maybe someone else has a clue already?

--- a/test/functional/rpcbind_test.py
+++ b/test/functional/rpcbind_test.py
@@ -47,9 +47,12 @@ class RPCBindTest(BitcoinTestFramework):
         at a non-localhost IP.
         '''
         self.log.info("Allow IP test for %s:%d" % (rpchost, rpcport))
-        base_args = ['-disablewallet', '-nolisten'] + ['-rpcallowip='+x for x in allow_ips]
+        node_args = \
+            ['-disablewallet', '-nolisten'] + \
+            ['-rpcallowip='+x for x in allow_ips] + \
+            ['-rpcbind='+addr for addr in ['127.0.0.1', "%s:%d" % (rpchost, rpcport)]] # Bind to localhost as well so start_nodes doesn't hang
         self.nodes[0].rpchost = None
-        self.start_nodes([base_args])
+        self.start_nodes([node_args])
         # connect to node through non-loopback interface
         node = get_rpc_proxy(rpc_url(get_datadir_path(self.options.tmpdir, 0), 0, "%s:%d" % (rpchost, rpcport)), 0, coveragedir=self.options.coveragedir)
         node.getnetworkinfo()
@@ -100,9 +103,9 @@ class RPCBindTest(BitcoinTestFramework):
             # check default without rpcallowip (IPv4 and IPv6 localhost)
             self.run_bind_test(None, '127.0.0.1', [],
                 [('127.0.0.1', self.defaultport), ('::1', self.defaultport)])
-            # check default with rpcallowip (IPv6 any)
+            # check default with rpcallowip (IPv4 and IPv6 localhost)
             self.run_bind_test(['127.0.0.1'], '127.0.0.1', [],
-                [('::0', self.defaultport)])
+                [('127.0.0.1', self.defaultport), ('::1', self.defaultport)])
             # check only IPv6 localhost (explicit)
             self.run_bind_test(['[::1]'], '[::1]', ['[::1]'],
                 [('::1', self.defaultport)])

--- a/test/functional/rpcbind_test.py
+++ b/test/functional/rpcbind_test.py
@@ -4,7 +4,6 @@
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 """Test running dashd with the -rpcbind and -rpcallowip options."""
 
-import socket
 import sys
 
 from test_framework.test_framework import BitcoinTestFramework, SkipTest
@@ -18,6 +17,11 @@ class RPCBindTest(BitcoinTestFramework):
 
     def setup_network(self):
         self.add_nodes(self.num_nodes, None)
+
+    def add_options(self, parser):
+        parser.add_option("--ipv4", action='store_true', dest="run_ipv4", help="Run ipv4 tests only", default=False)
+        parser.add_option("--ipv6", action='store_true', dest="run_ipv6", help="Run ipv6 tests only", default=False)
+        parser.add_option("--nonloopback", action='store_true', dest="run_nonloopback", help="Run non-loopback tests only", default=False)
 
     def run_bind_test(self, allow_ips, connect_to, addresses, expected):
         '''
@@ -53,55 +57,69 @@ class RPCBindTest(BitcoinTestFramework):
 
     def run_test(self):
         # due to OS-specific network stats queries, this test works only on Linux
+        if sum([self.options.run_ipv4, self.options.run_ipv6, self.options.run_nonloopback]) > 1:
+            raise AssertionError("Only one of --ipv4, --ipv6 and --nonloopback can be set")
+
+        self.log.info("Check for linux")
         if not sys.platform.startswith('linux'):
-            raise SkipTest("This test can only be run on Linux.")
-        # find the first non-loopback interface for testing
-        non_loopback_ip = None
+            raise SkipTest("This test can only be run on linux.")
+
+        self.log.info("Check for ipv6")
+        have_ipv6 = test_ipv6_local()
+        if not have_ipv6 and not self.options.run_ipv4:
+            raise SkipTest("This test requires ipv6 support.")
+
+        self.log.info("Check for non-loopback interface")
+        self.non_loopback_ip = None
         for name,ip in all_interfaces():
             if ip != '127.0.0.1':
-                non_loopback_ip = ip
+                self.non_loopback_ip = ip
                 break
-        if non_loopback_ip is None:
-            raise SkipTest("This test requires at least one non-loopback IPv4 interface.")
-        try:
-            s = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
-            s.connect(("::1",1))
-            s.close
-        except OSError:
-            raise SkipTest("This test requires IPv6 support.")
+        if self.non_loopback_ip is None and self.options.run_nonloopback:
+            raise SkipTest("This test requires a non-loopback ip address.")
 
-        self.log.info("Using interface %s for testing" % non_loopback_ip)
+        self.defaultport = rpc_port(0)
 
-        defaultport = rpc_port(0)
+        if not self.options.run_nonloopback:
+            self._run_loopback_tests()
+        if not self.options.run_ipv4 and not self.options.run_ipv6:
+            self._run_nonloopback_tests()
 
-        # check default without rpcallowip (IPv4 and IPv6 localhost)
-        self.run_bind_test(None, '127.0.0.1', [],
-            [('127.0.0.1', defaultport), ('::1', defaultport)])
-        # check default with rpcallowip (IPv6 any)
-        self.run_bind_test(['127.0.0.1'], '127.0.0.1', [],
-            [('::0', defaultport)])
-        # check only IPv4 localhost (explicit)
-        self.run_bind_test(['127.0.0.1'], '127.0.0.1', ['127.0.0.1'],
-            [('127.0.0.1', defaultport)])
-        # check only IPv4 localhost (explicit) with alternative port
-        self.run_bind_test(['127.0.0.1'], '127.0.0.1:32171', ['127.0.0.1:32171'],
-            [('127.0.0.1', 32171)])
-        # check only IPv4 localhost (explicit) with multiple alternative ports on same host
-        self.run_bind_test(['127.0.0.1'], '127.0.0.1:32171', ['127.0.0.1:32171', '127.0.0.1:32172'],
-            [('127.0.0.1', 32171), ('127.0.0.1', 32172)])
-        # check only IPv6 localhost (explicit)
-        self.run_bind_test(['[::1]'], '[::1]', ['[::1]'],
-            [('::1', defaultport)])
-        # check both IPv4 and IPv6 localhost (explicit)
-        self.run_bind_test(['127.0.0.1'], '127.0.0.1', ['127.0.0.1', '[::1]'],
-            [('127.0.0.1', defaultport), ('::1', defaultport)])
+    def _run_loopback_tests(self):
+        if self.options.run_ipv4:
+            # check only IPv4 localhost (explicit)
+            self.run_bind_test(['127.0.0.1'], '127.0.0.1', ['127.0.0.1'],
+                [('127.0.0.1', self.defaultport)])
+            # check only IPv4 localhost (explicit) with alternative port
+            self.run_bind_test(['127.0.0.1'], '127.0.0.1:32171', ['127.0.0.1:32171'],
+                [('127.0.0.1', 32171)])
+            # check only IPv4 localhost (explicit) with multiple alternative ports on same host
+            self.run_bind_test(['127.0.0.1'], '127.0.0.1:32171', ['127.0.0.1:32171', '127.0.0.1:32172'],
+                [('127.0.0.1', 32171), ('127.0.0.1', 32172)])
+        else:
+            # check default without rpcallowip (IPv4 and IPv6 localhost)
+            self.run_bind_test(None, '127.0.0.1', [],
+                [('127.0.0.1', self.defaultport), ('::1', self.defaultport)])
+            # check default with rpcallowip (IPv6 any)
+            self.run_bind_test(['127.0.0.1'], '127.0.0.1', [],
+                [('::0', self.defaultport)])
+            # check only IPv6 localhost (explicit)
+            self.run_bind_test(['[::1]'], '[::1]', ['[::1]'],
+                [('::1', self.defaultport)])
+            # check both IPv4 and IPv6 localhost (explicit)
+            self.run_bind_test(['127.0.0.1'], '127.0.0.1', ['127.0.0.1', '[::1]'],
+                [('127.0.0.1', self.defaultport), ('::1', self.defaultport)])
+
+    def _run_nonloopback_tests(self):
+        self.log.info("Using interface %s for testing" % self.non_loopback_ip)
+
         # check only non-loopback interface
-        self.run_bind_test([non_loopback_ip], non_loopback_ip, [non_loopback_ip],
-            [(non_loopback_ip, defaultport)])
+        self.run_bind_test([self.non_loopback_ip], self.non_loopback_ip, [self.non_loopback_ip],
+            [(self.non_loopback_ip, self.defaultport)])
 
         # Check that with invalid rpcallowip, we are denied
-        self.run_allowip_test([non_loopback_ip], non_loopback_ip, defaultport)
-        assert_raises_rpc_error(-342, "non-JSON HTTP response with '403 Forbidden' from server", self.run_allowip_test, ['1.1.1.1'], non_loopback_ip, defaultport)
+        self.run_allowip_test([self.non_loopback_ip], self.non_loopback_ip, self.defaultport)
+        assert_raises_rpc_error(-342, "non-JSON HTTP response with '403 Forbidden' from server", self.run_allowip_test, ['1.1.1.1'], self.non_loopback_ip, self.defaultport)
 
 if __name__ == '__main__':
     RPCBindTest().main()

--- a/test/functional/test_runner.py
+++ b/test/functional/test_runner.py
@@ -131,6 +131,9 @@ BASE_SCRIPTS= [
     'signmessages.py',
     'nulldummy.py',
     'import-rescan.py',
+    'rpcbind_test.py --ipv4',
+    'rpcbind_test.py --ipv6',
+    'rpcbind_test.py --nonloopback',
     'mining.py',
     'rpcnamedargs.py',
     'listsinceblock.py',
@@ -172,7 +175,6 @@ EXTENDED_SCRIPTS = [
     'getblocktemplate_longpoll.py',  # FIXME: "socket.error: [Errno 54] Connection reset by peer" on my Mac, same as  https://github.com/bitcoin/bitcoin/issues/6651
     'p2p-timeouts.py',
     # vv Tests less than 60s vv
-    'rpcbind_test.py',
     # vv Tests less than 30s vv
     'assumevalid.py',
     'example_test.py',

